### ISSUE TITLE
[Feature] : Wav2VecTokenizer

### ISSUE THREAD

**Issue by Ierezell** [huggingface/tokenizers#667]
It's possible to export a Wav2VecForCTC Huggingface model to onnx and to use it in other programming stack. 

<details>
<summary>Code for exporting a Wav2Vec Model</summary>
<pre>
<code>
from transformers import Wav2Vec2Tokenizer, Wav2Vec2ForCTC
from datasets import load_dataset
import torch
import os
import shutil
import soundfile as sf

MODEL_NAME = "facebook/wav2vec2-base-960h"
MODEL_FOLDER = f"./asr/{MODEL_NAME.replace('/','_')}"
OUTPUT_PATH = f"{MODEL_FOLDER}/model.onnx"

if os.path.exists(MODEL_FOLDER):
    shutil.rmtree(MODEL_FOLDER)

os.makedirs(MODEL_FOLDER)

tokenizer = Wav2Vec2Tokenizer.from_pretrained(MODEL_NAME)
model = Wav2Vec2ForCTC.from_pretrained(MODEL_NAME)
model = model.eval()

def map_to_array(batch):
    speech, _ = sf.read(batch["file"])
    batch["speech"] = speech
    return batch

ds = load_dataset("patrickvonplaten/librispeech_asr_dummy", "clean", split="validation")
ds = ds.map(map_to_array)

dummy_input = tokenizer(ds["speech"][:2], return_tensors="pt",
                        padding="longest").input_values

torch.onnx.export(model, dummy_input, OUTPUT_PATH, verbose=True,
                  input_names=["input"], output_names=["output"],
                  dynamic_axes={'input': {1: 'sequence'}})

tokenizer.save_pretrained(MODEL_FOLDER)
</code>
</pre>
</details>

<details>
<summary>Code for using the Wav2Vec model in nodejs</summary>
<pre>
<code>
import * as tf from '@tensorflow/tfjs-node';
import fs from 'fs';
import * as ort from "onnxruntime";
import path from 'path';

export class ASR
{
  private _session?: ort.InferenceSession
  private tokToId: { [key: string]: number }
  private idToTok: { [key: number]: string } = {}

  constructor(private modelPath: string)
  {
    this.tokToId = JSON.parse(fs.readFileSync(path.join(this.modelPath, 'vocab.json'), 'utf-8'))
    for (const [key, value] of Object.entries(this.tokToId)) { this.idToTok[value] = key }
  }

  async load() { this._session = await ort.InferenceSession.create(path.join(this.modelPath, 'model.onnx')) }

  decodeIds(ids: number[]): string
  {
    return ids.slice(0, Math.floor(ids.length / 2)).map(i => this.idToTok[i]).filter(t => t !== '<pad>').join("_")
  }

  async asr(vector: number[]): Promise<string>
  {
    const input_ids = new ort.Tensor("float32", vector.concat(vector), [2, vector.length])
    const results = await this._session!.run({ input: input_ids });
    const outTens = tf.tensor(results.output.data as Float32Array).reshape(results.output.dims as number[])
    const ids = tf.argMax(outTens, -1)
    return this.decodeIds(Array.from(ids.dataSync()))
  }
}

const test = async () =>
{
  const asr = new ASR(path.join(__dirname, '../models/asr/facebook_wav2vec2-base-960h'))
  const data = JSON.parse(fs.readFileSync(path.join(__dirname, '../models/asr/test-mic.json'), 'utf-8'))
  await asr.load()
  const res = await asr.asr(data.data)
  console.log(res)
}
test().then(() => { console.log("Done") })

// Gives : I|||THHINK|||IIT'SS|||A|||BEEAAUTIFUULL|||DAY|||TO|DAY|||
</code>
</pre>
</details>

However, I had to make a handmade ids decoder for it. It's working and can be used quite nicely but I was hopping that this kind of tokenizers would be part of the library and usable in nodejs. 

Ps : By using the code above, the output contain non currated words (multiple same letters) do you know how to fix that ? 

Thanks for this nice library ! 
Have a great day.

**Comment by ThomAub** [comment]
Hello, 

I think you might want to take a look at the class [Wav2Vec2CTCTokenizer](https://huggingface.co/transformers/model_doc/wav2vec2.html#transformers.Wav2Vec2CTCTokenizer). 

You are using a `Wav2Vec2ForCTC` which is a Wav2Vec2 Model with a language modeling head on top for Connectionist Temporal Classification (CTC). 

You can try to port the [convert_tokens_to_string](https://github.com/huggingface/transformers/blob/76116f479b1e115e524f587e99471279df5c49b0/src/transformers/models/wav2vec2/tokenization_wav2vec2.py#L204) in nodejs. 

If I have some time I can try to port both `Wav2Vec2(CTC)Tokenizer` to the rust lib but might take some time and note sure if this will be wanted by the maintainers

**Comment by Ierezell** [comment]
Hi @ThomAub, 

Thanks a lot for the reply. 

Indeed I should've RTFM better I haven't seen the CTC Tokenizer. 

However as you said, in python I should call the `convert_tokens_to_string` which will have the logic I want to group the letters together again (which will not be implemented in NodeJs). 

I will rewrite the function (which is quite small) and tell you about it  ! 

Thanks again, 
Have a great day

**Comment by ThomAub** [comment]
Hello @Ierezell, 

Glad it could help. 

I think it will be up to maintainers to see if those custom Tokenizer should be only in `transformers` or if a `fast` version could be also added here. 
Maybe ping @patrickvonplaten for a specific answer on `Wav2Vec2(CTC)Tokenizer`

**Comment by patrickvonplaten** [comment]
It's not super high priority, but we could definitely implement a Wav2Vec2CTCTokenizer in Rust -> PRs are always welcome :-)

**Comment by ThomAub** [comment]
Ok I might take a shot at it.

**Comment by Narsil** [comment]
This repo does not really have any concept of model specific tokenizers. They should rely mostly on `pre_processors` , `normalizers` and `models` (BPE, Unigram, WordLevel, etc..).

Wav2Vec2 should be a almost ready with a selection of those. 
Reading through Wav2Vec2Tokenizer, it really seems like most of the elements are already present within   `tokenizers` to get a Fast Tokenizer.

- model: WordLevel.
- normalizers/pre_tokenizers (not really necessary, as this tokenizer is mostly used as a decoder, not really as an encoder)
-  decoders: This is the one we should focus on. It's mostly a deduplicator of consecutive tokens.

@ThomAub if you want to take a stab at it, focus start by just implementing that (basically a function that takes -> A A A B A A and gives out A B A). The `clean_up_tokenization` part is already within  `decoders.WordPiece`.
Add a few rust tests, and it should be enough for a first review.

Implementing the rest of the scaffolding within Python/Rust/NodeJS is slightly more convoluted for a beginner but should be roughly trivial for us to help you with.

Feel free to ask any questions when writing the PR

**Comment by ThomAub** [comment]
Hello, 

thanks for the additional explanation ! 

For my application I ended up duplicating the `decoders.WordPiece` with this extra functionality (mostly calling `itertools.groupby`) and the two functions for mapping ID<>Tokens. 

How do you want this function integrated in your crate ? 

If I understand correclty the `TokenizerBuilder` currently only take an `Option<Decoder>` so there is no easy way to chain them. 

Would A separate module in `decoders` for `CTC`like input with the `CTC` struct implementing the `Decoder` trait be ok for you ? Also thinking about refactoring the `cleanup` part in `WordPiece` to use it as well an limiting code duplication.

**Comment by Narsil** [comment]
If we don't have a way to chain  `decoders` we should. We called them `Sequence` in `pre_tokenizers` for instance.

That should enable us to do the  `CTC` part only. When you start the PR for CTC I will dedicate a bit of time to make the rest of the plumbing work (`Sequence` and the bindings.)

**Comment by ThomAub** [comment]
Oh I just saw your comment. This is the current state of the PR. 
Let me know here or on the PR if you want me to change something :) 
With the help of something like `sequence` I would remove the part of calling `WordPiece.cleanup`

**Comment by Narsil** [comment]
@ThomAub , looks like a great PR, I replied a bit fast yesterday, chaining them will be much trickier, and maybe not as useful.

What you did is a great starting point.

**Comment by n1t0** [comment]
Added in https://github.com/huggingface/tokenizers/pull/693, will be included in the next release.
